webui/react/src/components/FilterForm/components/FilterFormStore.ts

```diff
@@ -128,152 +131,125 @@ export class FilterFormStore {
     );
   }
 
-  #getFormById(filterGroup: FormGroup, id: string): FormField | FormGroup | undefined {
-    const traverse = (form: FormGroup | FormField): FormGroup | FormField | undefined => {
-      if (form.id === id) {
-        return form;
-      }
-      if (form.kind === FormKind.Group && form.children.length === 0) {
-        return undefined;
-      }
-
-      if (form.kind === FormKind.Group) {
-        for (const child of form.children) {
-          const ans = traverse(child);
-          if (ans) {
-            return ans;
+  #updateForm(id: string, patch: PatchFunc<FormField | FormGroup>): void {
+    this.#formset.update((loadableFilterSet) => {
+      return Loadable.map(loadableFilterSet, (filterSet) => {
+        // keep updates to a minimum -- ids should be unique, so all following traversal should have no effect
```

Comment:
Couldn't you use a `for ... in` construct to actually short-circuit these loops?

Reply:
We'd still need to keep track of whether or not the change occurred and traverse the rest of the array (albeit not with the function), i think. If you imagine something like
```tsx
const newChildren = []
for(const child of entity.children) {
  const newChild = traverse(child)
  newChildren.push(newChild)
  if (newChild !== child) {
    break
  }
}
```
the `newChildren` array would still need to contain any children not visited. Furthermore, because we don't want to return a new array if no children were affected, we'd need to either do a deep equality check on the new and old children arrays, or check if the for loop broke.

Reply:
Right, got it.